web3/contract.py

```diff
@@ -1397,12 +1397,15 @@ def call_contract_function(
 
 
 def parse_block_identifier(web3, block_identifier):
-    last_block = web3.eth.getBlock('latest').number
-    if isinstance(block_identifier, int) and abs(block_identifier) <= last_block:
-        if block_identifier >= 0:
-            return block_identifier
+    if isinstance(block_identifier, int):
+        last_block = web3.eth.getBlock('latest').number
+        if abs(block_identifier) <= last_block:
+            if block_identifier >= 0:
+                return block_identifier
+            else:
+                return last_block + block_identifier + 1
```

Comment:
If the chain is 3 blocks long (so the latest number is 2), then -3 should be a valid identifier. One way to fix that edge case, and hopefully in a clean way to read, is:

```py
last_block = web3.eth.getBlock('latest').number

if block_identifier >= 0:
  block_num = block_identifier
else:
  block_num = last_block + block_identifier + 1

if 0 <= block_num <= last_block:
  return block_num
else:
  raise BlockNumberOutofRange
```

This is just long enough that it could go into its own method, like `parse_block_identifier_int()` and keep `parse_block_identifier` as a super simple dispatch method.